telemetry-setup/action.yml

```diff
@@ -0,0 +1,48 @@
+name: "GHA telemetry info"
+description: Returns identifier metadata that is useful in OpenTelemetry
+outputs:
+  trace_id:
+    description: Global ID; one per repo/run combination. Shared among all jobs in the run.
+    value: ${{ steps.trace-id.outputs.TRACE_ID }}
+  span_id:
+    description: |
+      Trace id combined with job name - this is the span that wraps the whole job.
+      There will be multiple spans per top-level trace ID.
+    value: ${{ steps.span-id.outputs.SPAN_ID }}
+  job_info_json:
+    description: Full job info, including steps with their status and start/complete times
+    value: ${{ steps.job-info.outputs.JOB_INFO_JSON }}
+  job_name:
+    description: Human-readable identifier for this job (often a description of a matrix step)
+    value: ${{ steps.job-info.outputs.JOB_NAME }}
+
+runs:
+  using: composite
+  steps:
+    - name: Generate Trace ID
+      id: trace-id
+      shell: bash
+      run: |
+        echo TRACE_ID=$(echo -n "${GITHUB_REPOSITORY}${GITHUB_RUN_ID}${GITHUB_RUN_ATTEMPT}" | sha256sum | cut -d ' ' -f 1) | tee --append "${GITHUB_ENV}" "${GITHUB_OUTPUT}"
+
+    - name: Get GitHub Job Info
+      id: job-info
+      shell: bash
+      env:
+        GH_TOKEN: ${{ github.token }}
+      # We know what the run ID is, but we don't know which specific job we're being run from.
+      # https://github.com/orgs/community/discussions/8945
+      run: |
+        JOBS="$(gh api \
+          -H "Accept: application/vnd.github+json" \
+          -H "X-GitHub-Api-Version: 2022-11-28" \
+          /repos/${GITHUB_REPOSITORY}/actions/runs/${GITHUB_RUN_ID}/jobs)"
+        JOB_INFO_JSON="$(echo "${JOBS}" | jq -c --arg runner_name "${RUNNER_NAME}" --arg attempt "${GITHUB_RUN_ATTEMPT}" '.jobs[] | select( .runner_name == $runner_name and .run_attempt == ($attempt |tonumber) )')"
+        echo JOB_INFO_JSON="${JOB_INFO_JSON}" | tee --append "${GITHUB_ENV}" "${GITHUB_OUTPUT}"
+        echo JOB_NAME="$(echo -n "${JOB_INFO_JSON}" | jq -c '.name')" | tee --append "${GITHUB_ENV}" "${GITHUB_OUTPUT}"
+
+    - name: Generate Job Span ID
+      id: span-id
+      shell: bash
+      run: |
+        echo SPAN_ID=$(echo -n "${GITHUB_RUN_ID}${GITHUB_RUN_ATTEMPT}${{ steps.job-info.outputs.JOB_NAME }}" | sha256sum | cut -d ' ' -f 1) | tee --append "${GITHUB_ENV}" "${GITHUB_OUTPUT}"
```

Comment:
As part of GitHub's best practices outlined below, I think we should move the `${{ steps.job-info.outputs.JOB_NAME }}` expression to an environment variable to mitigate the possibility of script injection attacks.

- https://docs.github.com/en/actions/security-for-github-actions/security-guides/security-hardening-for-github-actions#using-an-intermediate-environment-variable

Reply:
Done. I'm not super clear on how this could be used to engineer an attack, but I guess if you theoretically made a job name something like ````$(rm -rf /)````, it might be evaluated by the shell here.